python/revng/internal/cli/_commands/generate_migrations/__init__.py

```diff
@@ -0,0 +1,239 @@
+        try:
+            jsonschema.validate(instance=unstructured_schema, schema=metaschema)
+        except jsonschema.ValidationError as error:
+            raise GenerateMigrationException(f"Schema is invalid: {error.message}")
+
+        return self._restructure(unstructured_schema)
+
+    def _changed_keys(self, previous_properties: Dict, current_properties: Dict) -> List[str]:
+        result: List[str] = list(set(previous_properties) ^ set(current_properties))
+        for key in set(previous_properties) & set(current_properties):
+            if previous_properties[key] != current_properties[key]:
+                result.append(key)
+        return result
+
+    Change = (
+        Tuple[Literal["definition"], Literal["add"] | Literal["remove"], str]
+        | Tuple[Literal["definition"], Literal["edit"], str, str]
+        | Tuple[Literal["sub_definition"], Literal["add"] | Literal["remove"], str, str]
+        | Tuple[Literal["sub_definition"], Literal["edit"], str, str, str]
+    )
+
+    def _single_definition_changes(
+        self, definition_name: str, old_sub_definitions: Dict, new_sub_definitions: Dict
+    ) -> List[Change]:
+        result: List[GenerateMigrationCommand.Change] = []
+        for name in old_sub_definitions:
+            if name in new_sub_definitions:
+                for key in self._changed_keys(old_sub_definitions[name], new_sub_definitions[name]):
+                    result.append(("sub_definition", "edit", definition_name, name, key))
+            else:
+                result.append(("sub_definition", "remove", definition_name, name))
+        for name in new_sub_definitions:
+            if name not in old_sub_definitions:
+                result.append(("sub_definition", "add", definition_name, name))
+
+        return result
+
+    def _changes(self, old_schema: Dict, new_schema: Dict) -> List[Tuple]:
+        result: List[GenerateMigrationCommand.Change] = []
+        for name in old_schema["definitions"]:
+            if name in new_schema["definitions"]:
+                old_properties, old_sub_definitions = old_schema["definitions"][name]
+                new_properties, new_sub_definitions = new_schema["definitions"][name]
+
+                for key in self._changed_keys(old_properties, new_properties):
+                    result.append(("definition", "edit", name, key))
+
+                result.extend(
+                    self._single_definition_changes(
+                        name,
+                        old_sub_definitions,
+                        new_sub_definitions,
+                    )
+                )
+            else:
+                result.append(("definition", "remove", name))
+        for name in new_schema["definitions"]:
+            if name not in old_schema["definitions"]:
+                result.append(("definition", "add", name))
+
+        return result
+
+    def _todo_comment(self, change: Change):
```

Comment:
The usage of jinja we're doing here is quite minimal.
I'd suggest to either inline the jinja template or move the `_todo_comment` logic to jinja.